nav2_waypoint_follower/include/nav2_waypoint_follower/plugins/photo_at_waypoint.hpp

```diff
@@ -0,0 +1,125 @@
+
+#ifndef NAV2_WAYPOINT_FOLLOWER__PLUGINS__PHOTO_AT_WAYPOINT_HPP_
+#define NAV2_WAYPOINT_FOLLOWER__PLUGINS__PHOTO_AT_WAYPOINT_HPP_
+
+#include <experimental/filesystem>
+#include <mutex>
+#include <string>
+#include <exception>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+
+#include "sensor_msgs/msg/image.hpp"
+#include "nav2_core/waypoint_task_executor.hpp"
+#include "opencv4/opencv2/core.hpp"
+#include "opencv4/opencv2/opencv.hpp"
+#include "cv_bridge/cv_bridge.h"
+#include "image_transport/image_transport.hpp"
+
+
+namespace nav2_waypoint_follower
+{
+
+class PhotoAtWaypoint : public nav2_core::WaypointTaskExecutor
+{
+public:
+  /**
+  * @brief Construct a new Photo At Waypoint object
+  *
+  */
+  PhotoAtWaypoint();
+
+  /**
+   * @brief Destroy the Photo At Waypoint object
+   *
+   */
+  ~PhotoAtWaypoint();
+
+  /**
+   * @brief declares and loads parameters used
+   *
+   * @param parent parent node that plugin will be created within
+   * @param plugin_name should be provided in nav2_params.yaml==> waypoint_follower
+   */
+  void initialize(
+    const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+    const std::string & plugin_name);
+
+
+  /**
+   * @brief Override this to define the body of your task that you would like to execute once the robot arrived to waypoint
+   *
+   * @param curr_pose current pose of the robot
+   * @param curr_waypoint_index current waypoint, that robot just arrived
+   * @return true if task execution was successful
+   * @return false if task execution failed
+   */
+  bool processAtWaypoint(
+    const geometry_msgs::msg::PoseStamped & curr_pose, const int & curr_waypoint_index);
+
+  /**
+   * @brief
+   *
+   * @param msg
+   */
+  void imageCallback(const sensor_msgs::msg::Image::SharedPtr msg);
+
+  /**
+   * @brief given a shared pointer to sensor::msg::Image type, make a deep copy to inputted cv Mat
+   *
+   * @param msg
+   * @param mat
+   */
+  static void deepCopyMsg2Mat(const sensor_msgs::msg::Image::SharedPtr & msg, cv::Mat & mat);
+
+  /**
+   * @brief given encoding string , determine corresponding CV format
+   *
+   * @param encoding
+   * @return int
+   */
+  static int encoding2mat_type(const std::string & encoding);
+
+  /**
+   * @brief given CV type encoding return corresponding sensor_msgs::msg::Image::Encoding string
+   *
+   * @param mat_type
+   * @return std::string
+   */
+  static std::string mat_type2encoding(int mat_type);
+
+protected:
+  // to ensure safety when accessing global var curr_frame_
+  std::mutex global_mutex_;
+  // the taken photos will be saved under this directory
+  std::string directory_to_save_images_;
+  // .png ? .jpg ? or some other well known format
+  std::string image_format_;
+  // the topic to subscribe in order capture a frame
+  std::string image_topic_;
+  // whether plugin is enabled
+  bool is_enabled_;
+  // current frame;
+  sensor_msgs::msg::Image curr_frame_msg_;
```

Comment:
Store the shared pointer, that's why the callback gives you a shared pointer, don't put this on the stack